torch/csrc/distributed/rpc/request_callback_impl.cpp

```diff
@@ -358,52 +347,34 @@ void RequestCallbackImpl::processPythonRemoteCall(
     ctx.addForkOfOwner(rrefId, forkId);
   }
 
-  if (uprc.isAsyncExecution()) {
-    try {
-      processAsyncExecution(
-          uprc.pythonUdf(),
-          messageId,
-          responseFuture,
-          [ownerRRef, rrefId, forkId](
-              const py::object& result,
-              const int64_t messageId,
-              PythonRpcHandler& /* unused */,
-              const std::shared_ptr<JitFuture>& responseFuture) {
-            IValue py_ivalue = jit::toIValue(result, PyObjectType::get());
-
-            py::gil_scoped_release release;
-            ownerRRef->setValue(std::move(py_ivalue));
-            auto m = RemoteRet(rrefId, forkId).toMessage();
-            m.setId(messageId);
-            responseFuture->markCompleted(
-                IValue(c10::make_intrusive<Message>(std::move(m))));
-          });
-    } catch (std::exception& e) {
-      ownerRRef->setError(std::current_exception());
-      auto m = RemoteRet(rrefId, forkId).toMessage();
-      m.setId(messageId);
-      responseFuture->markCompleted(
-          IValue(c10::make_intrusive<Message>(std::move(m))));
-    }
-  } else {
-    IValue py_ivalue;
-    try {
-      {
-        py::gil_scoped_acquire acquire;
-        py_ivalue = jit::toIValue(
-            pythonRpcHandler.runPythonUdf(uprc.pythonUdf()),
-            PyObjectType::get());
-      }
-      ownerRRef->setValue(std::move(py_ivalue));
-    } catch (py::error_already_set& e) {
-      // py::error_already_set requires GIL to destruct, take special care.
-      ownerRRef->setError(std::current_exception());
-      py::gil_scoped_acquire acquire;
-      e.restore();
-      PyErr_Clear();
-    } catch (std::exception& e) {
-      ownerRRef->setError(std::current_exception());
-    }
+  try {
+    processPythonAsyncExecution(
+        uprc.pythonUdf(),
+        messageId,
+        responseFuture,
+        uprc.isAsyncExecution(),
+        [ownerRRef, rrefId, forkId, markComplete](
+            const py::object& result,
+            const int64_t messageId,
+            PythonRpcHandler& /* unused */,
+            const std::shared_ptr<JitFuture>& responseFuture) {
+          IValue py_ivalue = jit::toIValue(result, PyObjectType::get());
+
+          py::gil_scoped_release release;
```

Comment:
Unrelated and does not need to be addressed in this diff, but feel like this could be challenging to reason about. We acquire GIL before call to `postProcessing`, and then release it here, and then it's re-acquired back in the original scope, I think. Should we at least have a DCHECK() here that the GIL is held?